Getmail service

romanofski opened the pull request:
This is pull request #290 with addressed code review comments.

teto commented on modules/services/getmail.nix:
```diff
@@ -0,0 +1,193 @@
+    options = {
+      type = mkOption {
+        type = types.enum [
+          "SimplePOP3Retriever"
```
This submodule should not be necessary;  the config is already available in modules/accounts/email.nix: you get the username, the passwrod, the ssl support for imap and pop from there.

romanofski replied:
Many thanks for the pointer(s). Will adjust and update.

teto commented on modules/services/getmail.nix:
```diff
@@ -0,0 +1,193 @@
+          "Maildir"
+        ];
+        default = "Maildir";
+        description = "Destination type.";
```
likewise.

romanofski commented:
I'm chiping away on this one. Since the first pull request, home-manager seemed to have changed quite a bit when it comes to re-usability of email related information. I've started to move `getmail` into `programs` instead of leaving it as a service. Reasoning being is that there are all the other email related modules are there too: consistency.

Due to the fact that we're re-using most of the information from the general `emails` module, parts of the options in the `getmail` module are gone. Furthermore I realized that the module now needs to support more than one account, which means I'll need to write more than one configuration.

teto commented:
Not sure I get the meaning of `chiping away` (it gets translated in many different ways, including "reducing" ?). If getmail doesn't support multiple accounts, you could generate the config for the primary one `(filter (a: a.primary)  (attrValues cfg.accounts));` and merge that first.
Multiaccount can arrive later by you or someone else.

romanofski commented:
@teto cheers for the pointer. That will make life easier.

I guess it would have been more comprehensible if I would have not made a spelling error. So the correct way to write it is: chipping away meaning to make slow progress :)

romanofski commented:
Updated and looking for feedback.

 A couple of changes:
* Moved the module from services into programs
* Added additional `getmail-accounts` configuration, since not everything is exposed through email-accounts. Please let me know if I've overlooked something important.
* Also added tests, but am a bit puzzled how to run them. I see that `.travis.yml` uses a nix-shell for this. I'm not sure if that is the same way how I can run the tests locally? In other words: how can I execute the tests :)

Please don't merge yet. I'd like to run this setup for a couple of days to see that I haven't missed something important. I guess the tests are setup to fail for now which would prevent merging anyway.

rycee commented:
@romanofski The command in the Travis CI file should work just fine locally as well, just go to the checked out home-manager directory and run
```console
$ nix-shell tests -A run.all
```

Thanks for working on the module, great stuff! And thanks to @teto for reviewing 🙂 I'll try to also have a look on the weekend.

romanofski commented:
> @romanofski The command in the Travis CI file should work just fine [...]
> 

Aaah cheers. Yes that's helpful.


> Thanks for working on the module, great stuff! And thanks to @teto for reviewing slightly_smiling_face I'll try to also have a look on the weekend.

No worries. Thanks for the great software in the first place :)

romanofski commented:
Updated. The test is very simple, but hopefully will catch the worst regressions. The service uses a timer at the moment and I plan to create another pull request to allow the option of supporting `IDLE` mailboxes which would make the timer superfluous.
Let me know if there are things which can be done better. I'm more of a nix n00b.

romanofski commented:
I've updated the docstrings a bit. Just because I noticed that if someone wouldn't know about getmail, the documentation like:
```
programs.getmail.enable
           Whether to enable Enable getmail.
```
wouldn't really tell me about what `getmail` actually *is*. So I've amended some of the docstrings to:
```
programs.getmail.enable
           Whether to enable getmail, a mail retriever to get mail from your mail accounts.
```
to be hopefully more descriptive.

teto commented:
For mbsync, there is a dinstinction between the `service.getmail` that generates the systemd units and the `programs.getmail` that generates the config. I don't know if @rycee  wants to stick the model or change it to something else like  `program.runService = true;` ; it's a bit confusing to have both as top-level sets IMO. 

Also as I had forgotten to set getmail.enable in my mail accounts, I ended up with this unit (ie without any account target):
```
[Service]
ExecStart=/nix/store/l0b9ckflvn3jfqf0ljrksd8ikmk3q65g-getmail-5.13/bin/getmail 

[Unit]
Description=getmail email fetcher

```
The module shouldn't generate a unit if no account is configured. 

NB: I am not convinced by the new dostring, looks to me like "getmail: a program to get mail " xD The program name looks self explanatory.

romanofski commented:
@teto many thanks for the feedback and testing. I'll have a look.

romanofski commented:
Updated. I've added another assertion. Couldn't come up with anything better :)

teto commented on modules/programs/getmail-accounts.nix:
```diff
@@ -0,0 +1,46 @@
+
+    mailboxes = mkOption {
+      type = types.listOf types.str;
+      default = [];
```
does getmail mandate to set mailboxes ? would it make sense to set ["INBOX"] as default ?

teto commented on modules/programs/getmail.nix:
```diff
@@ -0,0 +1,115 @@
+  {
+    options = {
+      programs.getmail = {
+        enable = mkEnableOption "getmail, a mail retriever to get mail from your mail accounts";
```
I feel like "getmail" is fine but won't argue for it :p

teto commented:
thanks for the quick update.

I hit
```
error: while evaluating the attribute 'activationPackage' at /home/teto/home-manager/modules/default.nix:46:3:
while evaluating 'showWarnings' at /home/teto/home-manager/modules/default.nix:16:18, called from /home/teto/home-manager/modules/default.nix:31:12:
while evaluating 'foldr' at /home/teto/nixpkgs/lib/lists.nix:39:20, called from /home/teto/home-manager/modules/default.nix:20:7:
while evaluating 'fold'' at /home/teto/nixpkgs/lib/lists.nix:42:15, called from /home/teto/nixpkgs/lib/lists.nix:46:8:

Failed assertions:
- getmail: expected account gmail mailboxes should be non-empty
```
which made me wonder if it makes sens to have a  default ? like is it mandated by getmail ? also I think the message could be improved to sthg like  `account [gmail]  getmail.mailboxes should be non-empty`.

Btw it seems like there is no way to disable the service ? I think you should split the service from the progran, you can look as mbsync has both a program/mbsync and services/mbsync.

Sorry for the nitpicks. It's looking good. Hopefully we can merge next update (tests are a great addition).

teto commented:
thanks for the quick update.

I hit
```
error: while evaluating the attribute 'activationPackage' at /home/teto/home-manager/modules/default.nix:46:3:
while evaluating 'showWarnings' at /home/teto/home-manager/modules/default.nix:16:18, called from /home/teto/home-manager/modules/default.nix:31:12:
while evaluating 'foldr' at /home/teto/nixpkgs/lib/lists.nix:39:20, called from /home/teto/home-manager/modules/default.nix:20:7:
while evaluating 'fold'' at /home/teto/nixpkgs/lib/lists.nix:42:15, called from /home/teto/nixpkgs/lib/lists.nix:46:8:

Failed assertions:
- getmail: expected account gmail mailboxes should be non-empty
```
which made me wonder if it makes sens to have a  default ? like is it mandated by getmail ? also I think the message could be improved to sthg like  `account [gmail]  getmail.mailboxes should be non-empty`.

Btw it seems like there is no way to disable the service ? I think you should split the service from the progran, you can look as mbsync has both a program/mbsync and services/mbsync.

Sorry for the nitpicks. It's looking good. Hopefully we can merge next update (tests are a great addition).

romanofski commented:
@teto nah don't worry about the nitpicks. I appreciate them. Don't feel urged to green light something which is wonky.

In regards to a default: We could default to `"ALL"`. The problem with this could be users using GMail which would pull down not only your inbox, but every folder. That includes GMails `All Mail` folder, which could be huge. If we just default to `"Inbox"` it would cause errors if the default isn't named `"Inbox"`...

In regards to the service: do you have a case where you'd want to do that? E.g wouldn't I either enable getmail and therefore the service or not? The splitting itself is probably wise, since as I mentioned earlier I'd like to add `IDLE` support in a subsequent patch and this might make it easier. I'm onboard on the splitting the responsibilities, just would like to ask if you could elaborate more on why you'd want to disable the service. Maybe to run just the command which uses the config?

Cheers!

teto commented:
According to your explanation, it makes sense not to have a default so let's keep it that way.

I use mbsync but I run updates for my mail manually because I like it that way :D sometimes I need a VPN connection for my pro mail, sometimes I am on a metered connection so I don't want background services to consume my data. Other times I am on a bad connection and don't want mail fetching to take all the bandwidth etc.

Also I think nix has a future just to generate config (as an online service ?) so making a clear distinction between services may be helpful.

romanofski commented:
Updated. I thought the need to enable `getmail` at three different stages was a bit overkill and I changed `programs/getmail.nix` slightly. That is, `programs/getmail.nix` writes the configuration, if `getmail` is enabled in any one of the accounts, not when it is explicitly enabled via `programs.getmail.enabled = true`, since well, you do that already in your accounts.

This now has the `service.getmail` back in which basically enables the systemd service and timer to automatically invoke getmail.

Since this slightly breaks consistency with the rest (you always need to explicitly enable the programs), is this change acceptable or should I add the explicit `programs.getmail.enabled = true` back in?

teto commented:
Looks good. I'll let some time for rycee to comment, if he doesn't I'll merge it some time next week.

rycee commented on modules/modules.nix:
```diff
@@ -95,6 +96,7 @@ let
     (loadModule ./services/dunst.nix { })
     (loadModule ./services/emacs.nix { condition = hostPlatform.isLinux; })
     (loadModule ./services/flameshot.nix { })
+    (loadModule ./services/getmail.nix { })
```
Since the service part uses systemd it is limited to Linux, so can add `condition = hostPlatform.isLinux;` here.

romanofski replied:
Done.

rycee commented on modules/misc/news.nix:
```diff
@@ -1099,6 +1098,12 @@ in
+        message = ''
+          A new module is available: `services.getmail`.
+        '';
+}
```
Indentation.

romanofski replied:
Eeek. What happened there. Done.

rycee commented on modules/misc/news.nix:
```diff
@@ -1099,6 +1098,12 @@ in
           A new module is available: 'services.xsuspender'.
         '';
       }
+      {
```
Needs an empty line.

rycee commented on modules/misc/news.nix:
```diff
@@ -1065,7 +1065,6 @@ in
           A new module is available: 'programs.alacritty'.
         '';
       }
-
```
Should keep the empty line to keep a consistent format.

romanofski replied:
Makes sense. Done.

rycee commented on modules/programs/getmail-accounts.nix:
```diff
@@ -0,0 +1,46 @@
+      type = types.listOf types.str;
+      default = [];
+      example = ["INBOX" "INBOX.spam"];
+      description = "A list of mailboxes";
```
Descriptions should end in a `.` (the `mkEnableOption` argument is an exception since the `.` is added by the function).

romanofski replied:
Done.

rycee commented on modules/programs/getmail.nix:
```diff
@@ -0,0 +1,64 @@
+      retrieverType = if imap.tls.enable
+        then "SimpleIMAPSSLRetriever"
+        else "SimpleIMAPRetriever";
+      destination = if (getmail.destinationCommand != null)
```
The parentheses are unnecessary.

rycee replied:
This still needs resolving.

romanofski replied:
Eeek.. how could I have overlooked this. Cheers for the reminder!

romanofski replied:
Updated

rycee commented on modules/services/getmail.nix:
```diff
@@ -0,0 +1,61 @@
+{
+  options = {
+    services.getmail = {
+      enable = mkEnableOption "the getmail systemd service to automatically retrieve mail.";
```
`mkEnableOption` already adds a final `.` so no need to do it here.

rycee replied:
This still needs resolving.

rycee commented on modules/services/getmail.nix:
```diff
@@ -0,0 +1,61 @@
+    (attrValues config.accounts.email.accounts);
+
+  renderConfigFilepath = a: ".getmail/getmail${a.name}";
+  quoted = concatMapStrings (x: " --rcfile \${HOME}/${x}") (map renderConfigFilepath accounts);
```
Could simplify slightly:
```suggestion
  quoted = concatMapStringsSep " " (a: "--rcfile ${renderConfigFilepath a}")  accounts;
```

romanofski replied:
Ah yep. Cheers for this!

rycee commented on modules/services/getmail.nix:
```diff
@@ -0,0 +1,61 @@
+  accounts = filter (a: a.getmail.enable)
+    (attrValues config.accounts.email.accounts);
+
+  renderConfigFilepath = a: ".getmail/getmail${a.name}";
```
Could include `$HOME` here:
```suggestion
  renderConfigFilepath = a: "\${HOME}/.getmail/getmail${a.name}";
```

romanofski replied:
Yes that's better.

rycee commented on modules/services/getmail.nix:
```diff
@@ -0,0 +1,61 @@
+
+  renderConfigFilepath = a: ".getmail/getmail${a.name}";
+  quoted = concatMapStrings (x: " --rcfile \${HOME}/${x}") (map renderConfigFilepath accounts);
+  renderStartCmd = accounts: "${pkgs.getmail}/bin/getmail ${quoted}";
```
The `accounts` argument is unnecessary. In fact, could inline this directly into the `ExecStart` line below 🙂

romanofski replied:
Nice! Done.

rycee commented on modules/services/getmail.nix:
```diff
@@ -0,0 +1,61 @@
+        Description = "getmail email fetcher";
+      };
+      Service = {
+        ExecStart = renderStartCmd accounts;
```
```suggestion
        ExecStart = "${pkgs.getmail}/bin/getmail ${quoted}";
```

romanofski replied:
Heh true! Thank you!

rycee commented on modules/programs/getmail-accounts.nix:
```diff
@@ -0,0 +1,46 @@
+
+    mailboxes = mkOption {
+      type = types.listOf types.str;
+      default = [];
```
If the user must supply at least one value I would suggest to leave this empty and set the type to `types.nonEmptyListOf types.str`. Then the assertion in `programs/getmail.nix` should be unnecessary.

romanofski replied:
Aaah nice! I was wondering about such a type. Taking that, I now found the `lib/types.nix` module.

romanofski replied:
Removed the assertion. The only way I could make it fail now is by not specifying an `imap` submodule and/or some of it's attributes. Is there anything to guard against without expensive `assertions`? Or ignore that for now?

rycee replied:
The mbsync module contains
https://github.com/rycee/home-manager/blob/8991fe2e90a9b35a98a8b98f02c9de5425b11785/modules/programs/mbsync.nix#L147-L163
which you can copy if you want but I think you can skip it for now. I'll try to put together a more general solution to make these assertions as soon as possible.

romanofski replied:
Cheers @rycee. Should have looked into the module before commenting. I'll be happy to skip it for now and might add it with another patch for the `IDLE` support. I suppose there is no 100% prevention to stuff up the `imap` configuration to end up in an error.

teto commented:
@romanofski is this ready ? Could you address the mkEnableOption and rebase please ? then I think we can merge

romanofski commented:
@teto not sure how I missed this every.single.time. Very sorry. Thanks for the ping. Rebased and updated.

teto commented:
I think there might be sthg wrong with the generated path:
```
$ getmail -r ~/.getmail/getmailgmail                                                                                                                                                                 
Configuration error: configuration file /home/teto/.getmail/getmailgmail incorrect (path: specified directory "gmail" does not exist)
```
in the generated config, I see:
`path = gmail`

maybe the module needs to prepend some path as in mbsync's module ? 
`Inbox = "${maildir.absPath}/${folders.inbox}";`

romanofski commented:
@teto that seems to be the case. Thank you for finding that. Does your configuration set:
```nix
{
maildir.path = "gmail";
}
```
?

romanofski commented:
I realised that I should have used `maildir.absPath` instead of just `maildir.path`.  Updated.

teto commented:
it produces an absolute path but now I get `Configuration error: configuration file /home/teto/.getmail/getmailgmail incorrect (path: maildir must end with "/")`.

Also I have to set the path to the config file `getmail -r ~/.getmail/getmailgmail` while gmail is my primary account. Would be nice to have getmailrc generated for the primary account.
```
$ getmail                                                                                                                                                                                               
Error: configuration file /home/teto/.getmail/getmailrc does not exist
```

romanofski commented:
@teto I like the idea about setting the primary account to `getmailrc`, so I made the change. I'm not sure however about the trailing slash on the path. I checked and you could easily add it into your home configuration, although not very obvious. I feel adding a trailing slash nevertheless would be the worse solution unless there is a function I could use to add it if the name doesn't have a trailing slash?

teto commented on modules/programs/getmail.nix:
```diff
@@ -0,0 +1,59 @@
+        else
+        {
+          destinationType = "Maildir";
+          destinationPath = maildir.absPath;
```
shouldn't it be `destinationPath = maildir.absPath + "/${accout.name}/"; otherwise all accounts share the same destinationPath

teto commented:
I believe you can unconditionnally add the trailing slash since I don't imagine people naming their accounts with a slash (We can revisit later but would be nice to merge this). Forcing the user to add it breaks the abstraction for other modules ? seems like a straightforward fix.

romanofski commented:
@teto Ok, if you think that would be better I'll follow your advice. I've updated the patch set and changed it to concat the trailing segment unconditionally. Let me know what you think :)

romanofski commented:
Now I found a problem with the invokation of getmail and it's passing rcfiles. So don't merge yet please!

romanofski commented:
Alright now. I found out that getmail expects to have just the filename of the config to be passed in instead of the path. This is now fixed.

teto commented:
I wonder how you configure it ? because for me it still won't work:

`configuration error: configuration file /home/teto/.getmail/getmailrc incorrect (path: maildir subdirectory "/home/teto/maildir/gmail/cur" does not exist)`
with the per-account config:
```
    getmail={
      enable = true;
      mailboxes = ["INBOX" "Sent" "Work"];
    };
```
The path should be /home/teto/maildir/gmail/Inbox/cur or  /home/teto/maildir/gmail/Sent/cur . I don't use getmail can't we pass several paths ?

romanofski commented:
@teto I'm sorry. The mailboxes seem to be correctly configured. Checking the getmail help it seems to suggest if you choose this option it will not file the mail into sub maildirs:

> the requirements [for the path to the maildir] in a nutshell are: it must be a directory containing three writable subdirectories cur, new, and tmp, and they must all reside on the same filesystem. 

from http://pyropus.ca/software/getmail/configuration.html#destination-maildir

I personally use a local MTA for delivery like so:
```
        getmail = {
          enable = true;
          mailboxes = ["ALL"];
          destinationCommand = "${pkgs.maildrop}/bin/maildrop";
          delete = true;
        };
```
which then can be configured to delivery to whatever directories you please. That's why testing this part from view is always a bit tricky.

teto commented:
I am annoyed that it doesn't work with my inbox but it seems like mbsync config generated a maildir inbox per gmail folder which doesn't seem (straightforwardly) supported by getmail after looking at the doc.
I you have been responsive to the issues and it works for you, I suppose we could merge this and we can improve the module afterwards. @rycee what do you think ?

romanofski commented:
@teto It seems it's how getmail operates. I actually used to use fetchmail + procmail before and went to getmail + maildrop. I lost a few mails via procmail, but have not had the problem with maildrop. I can't remember any more if there was a specific reason as to why I chose getmail tho ...

teto commented:
yes I think gmail creating a different maildir per label is awkward, (another reason why I want to get rid of it). I plan to merge this in a few days waiting for comments, it should work with normal boxes, if someone has a way to make it work with gmail, then great but that can come later.

teto commented:
thanks for keeping up with the reviews. I hope you stick around in case the module raises further concerns :p

romanofski commented:
@teto many thanks. Hit me up if there is trouble.
